### PR TITLE
[Az.DataProtection] Added Grant permissions for all workloads

### PR DESCRIPTION
**Pull request by rishit-microsoft**
## Description

Added Grant permissions for all workloads

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - {Please put the link here}
    - [x] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

**Comment by VeryEarly** [comment]
there is an OOB going on for https://github.com/Azure/sdk-release-request/issues/2905
is this PR also need to be included in OOB release?

File: src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1 [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+          $SubscriptionName = "/subscriptions/" + $ResourceArray[2]
+          $vault = Get-AzDataProtectionBackupVault -VaultName $VaultName -ResourceGroupName $VaultResourceGroup -SubscriptionId $ResourceArray[2]
+          
+          $AllRoles =  Get-AzRoleAssignment -ObjectId $vault.IdentityPrincipalId
+
+          $MissingRolesInitially = $false
+
+          foreach($Permission in $manifest.keyVaultPermissions)
+          {              
+              Write-Host "We will attempt to automatically grant the foloowing access: `n
```
Comment (akshayneema) [inline src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1:55]: spelling error - 'following'

File: src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1 [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+                            1.Backup vault’s identity access on the Postgres server and the key vault `n
+                            2.'Allow All Azure services' under network connectivity in the Key vault `n"
+
+              $Choice = Read-Host "Press n/N to exit the grant permissions or press any other key to continue"
```
Comment (akshayneema) [inline src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1:59]: Is this the standard way of prompting input from user in PS? Also if the user wants to bypass these prompts is there any way?

Comment (akshayneema) [reply src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1:59]: Ideally there should be a parameter in the command which user can use to bypass all the prompts

File: src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1 [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+          if($DatasourceType -eq "Microsoft.Storage/storageAccounts"){$DatasourceType = "AzureBlob"}
+          elseif($DatasourceType -eq "Microsoft.Compute/disks"){$DatasourceType = "AzureDisk"}
+          elseif($DatasourceType -eq "Microsoft.DBforPostgreSQL/servers/databases"){$DatasourceType = "AzureDatabaseForPostgreSQL"}
+                 
+          $manifest = LoadManifest -DatasourceType $DatasourceType.ToString()
+
+          $ResourceArray = $DataSourceId.Split("/")
+          $ResourceRG = "/subscriptions/" + $ResourceArray[2] + "/resourceGroups/" + $ResourceArray[4]
+          $SubscriptionName = "/subscriptions/" + $ResourceArray[2]
```
Comment (akshayneema) [inline src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1:46]: instead of doing this can we have a helper functions like 

GetSubscriptionIdFromArmId and GetResourceGroupIdFromArmId. As all the resources have a subscription and a resource group, these functions will make code for reusable and readable.

File: src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1 [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+              $Choice = Read-Host "Press n/N to exit the grant permissions or press any other key to continue"
+
+              if($Choice.ToLower() -eq "n")
+              {
+                  Write-Host "Cancelling Grant permissions operation"
+                  return
+              }
+
+              $ServerRG = $ResourceArray[4]
```
Comment (akshayneema) [inline src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1:67]: For this too we can have a function like GetResourceGroupNameFromArmId

File: src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1 [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+                  Write-Host "KeyvaultId not provided .Please note that permissions on keyvault may not be successfully assigned."  
+                  break
+              }
+                            
+              $KeyvaultArray = $KeyvaultId.Split("/")
+              $KeyvaultRG = "/subscriptions/" + $KeyvaultArray[2] + "/resourceGroups/" + $KeyvaultArray[4]
+              $KeyvaultSubscriptionName = "/subscriptions/" + $KeyvaultArray[2]
+              $KeyvaultName = $KeyvaultArray[8]
+              $KeyVault = Get-AzKeyVault -VaultName $KeyvaultName
+
```
Comment (akshayneema) [inline src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1:84]: Here we need to add a check ensuring that the KeyVault provided as the command argument indeed has the secret provided as the part of backup instance JSON object.

File: src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1 [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+                  {
+                     AssignMissingRoles -ObjectId $vault.IdentityPrincipalId -Permission $Permission -ResourceScope $DatasourceId
+                  }
+              
+                  elseif($PermissionsScope -eq "ResourceGroup")
+                  {
+                     AssignMissingRoles -ObjectId $vault.IdentityPrincipalId -Permission $Permission -ResourceScope $ResourceRG
+                  }
+
+                  elseif($PermissionsScope -eq "Subscription")
```
Comment (akshayneema) [inline src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1:154]: This code can be shortened with use of a helper function that on providing the armid and the scope returns the corresponding scope id.

File: src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1 [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+                  $MissingRolesInitially = $true
+
+                  Write-Host "Assigning Required SnapshotRG Permissions"
+
+                  if($PermissionsScope -eq "Subscription")
+                  {
+                      AssignMissingRoles -ObjectId $vault.IdentityPrincipalId -Permission $Permission -ResourceScope $SubscriptionName
+                  }             
+                  else
+                  {
```
Comment (akshayneema) [inline src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1:187]: This code can be shortened with use of a helper function that on providing the armid and the scope returns the corresponding scope id.

**Comment by hiaga** [comment]
> there is an OOB going on for [Azure/sdk-release-request#2905](https://github.com/Azure/sdk-release-request/issues/2905) is this PR also need to be included in OOB release?

not needed

File: src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1 [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+          }
+               
+          foreach($Permission in $manifest.datasourcePermissions)
+          {
+              $CheckPermission = $AllRoles
+              | Where-Object { ($_.Scope -eq $DataSourceId -or $_.Scope -eq $ResourceRG -or  $_.Scope -eq $SubscriptionName) -and $_.RoleDefinitionName -eq $Permission}
+              
+              if($CheckPermission -ne $null)
+              {
+                  Write-Host "Required Resource Permissions Already Assigned."
```
Comment (hiaga) [inline src/DataProtection/custom/Cmdlets/Platform/Vault/Set-AzDataProtectionMSIPermission.ps1:180]: same, we can add permission name here

File: src/DataProtection/docs/Az.DataProtection.md [hunk 9]
```diff
@@ -101,6 +101,9 @@ Searches for Backup instances in Azure Resource Graph and retrieves the expected
 ### [Search-AzDataProtectionJobInAzGraph](Search-AzDataProtectionJobInAzGraph.md)
 Searches for Backup Jobs in Azure Resource Graph and retrieves the expected entries
 
+### [Set-AzDataProtectionMSIPermission](Set-AzDataProtectionMSIPermission.md)
+Initializes Backup instance Request object for configuring backup
```
Comment (hiaga) [inline src/DataProtection/docs/Az.DataProtection.md:105]: Change this description

File: src/DataProtection/docs/Az.DataProtection.md [hunk 10]
```diff
@@ -113,6 +116,18 @@ This operation will stop backups for backup instance
 ### [Sync-AzDataProtectionBackupInstance](Sync-AzDataProtectionBackupInstance.md)
 Sync backup instance again in case of failure\r\nThis action will retry last failed operation and will bring backup instance to valid state
 
+### [Test-AzDataProtectionBackupInstanceReadiness](Test-AzDataProtectionBackupInstanceReadiness.md)
+Validate whether adhoc backup will be successful or not
+
+### [Test-AzDataProtectionBackupInstanceRestore](Test-AzDataProtectionBackupInstanceRestore.md)
+Validates if Restore can be triggered for a DataSource
+
+### [Test-AzDataProtectionBackupVaultNameAvailability](Test-AzDataProtectionBackupVaultNameAvailability.md)
```
Comment (hiaga) [inline src/DataProtection/docs/Az.DataProtection.md:125]: remove Test-AzDataProtectionBackupVaultNameAvailability and Test-AzDataProtectionFeatureSupport

File: src/DataProtection/docs/Backup-AzDataProtectionBackupInstanceAdhoc.md [hunk 11]
```diff
@@ -34,27 +34,26 @@ Trigger adhoc backup
 
 ### Example 1: Backup a protected backup instance
 ```powershell
-PS C:\> $instance = Get-AzDataProtectionBackupInstance -SubscriptionId "xxxx-xxx-xxx" -ResourceGroupName "MyResourceGroup" -VaultName "MyVault"
-PS C:\> Backup-AzDataProtectionBackupInstanceAdhoc -BackupInstanceName $instance.Name -ResourceGroupName "MyResourceGroup" -SubscriptionId "xxxx-xxx-xxxx" -VaultName "MyVault" -BackupRuleOptionRuleName "BackupWeekly" -TriggerOptionRetentionTagOverride "Default"
```
Comment (hiaga) [inline src/DataProtection/docs/Backup-AzDataProtectionBackupInstanceAdhoc.md:38]: revert all such changes

File: src/DataProtection/docs/Find-AzDataProtectionRestorableTimeRange.md [hunk 12]
```diff
@@ -8,7 +8,7 @@ schema: 2.0.0
 # Find-AzDataProtectionRestorableTimeRange
```
Comment (hiaga) [inline src/DataProtection/docs/Find-AzDataProtectionRestorableTimeRange.md:9]: revert this file

File: src/DataProtection/docs/Initialize-AzDataProtectionBackupInstance.md [hunk 13]
```diff
@@ -139,6 +139,21 @@ Accept pipeline input: False
 Accept wildcard characters: False
 ```
 
+### -Snapshotrg
+Sanpshot Resource Group
```
Comment (hiaga) [inline src/DataProtection/docs/Initialize-AzDataProtectionBackupInstance.md:143]: regenerate the help after modifying the param name

File: src/DataProtection/docs/Set-AzDataProtectionMSIPermission.md [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+```powershell
+PS C:\> Set-AzDataProtectionMSIPermission -BackupInstance $instance -VaultResourceGroup "VaultRG"" -VaultName "Vaultname"" -PermissionsScope "ResourceGroup"
+
+```
+
+The above command is used to assign permissions to the backup vault "Vaultname" under resource group "VaultRG" at the "Rresource Group" scope of the disk.
+
+### Example 2: Grant Permissions for Azure Blobs
+```powershell
+PS C:\> Set-AzDataProtectionMSIPermission -BackupInstance $instance -VaultResourceGroup "VaultRG"" -VaultName "Vaultname"" -PermissionsScope "ResourceGroup"
```
Comment (hiaga) [inline src/DataProtection/docs/Set-AzDataProtectionMSIPermission.md:36]: you can make this example slightly different than example 1, for example at a different scope

File: src/DataProtection/docs/Test-AzDataProtectionBackupInstanceReadiness.md [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+```
+
+## DESCRIPTION
+Validate whether adhoc backup will be successful or not
+
+## EXAMPLES
+
+### Example 1: {{ Add title here }}
+```powershell
+{{ Add code here }}
```
Comment (hiaga) [inline src/DataProtection/docs/Test-AzDataProtectionBackupInstanceReadiness.md:50]: add examples

File: src/DataProtection/docs/Test-AzDataProtectionBackupInstanceRestore.md [hunk 16]
```diff
@@ -0,0 +1,354 @@
...
+```
+
+## DESCRIPTION
+Validates if Restore can be triggered for a DataSource
+
+## EXAMPLES
+
+### Example 1: {{ Add title here }}
+```powershell
+{{ Add code here }}
```
Comment (hiaga) [inline src/DataProtection/docs/Test-AzDataProtectionBackupInstanceRestore.md:54]: add example

File: src/DataProtection/docs/Test-AzDataProtectionBackupVaultNameAvailability.md [hunk 17]
```diff
@@ -0,0 +1,265 @@
+---
+external help file:
+Module Name: Az.DataProtection
+online version: https://docs.microsoft.com/powershell/module/az.dataprotection/test-azdataprotectionbackupvaultnameavailability
+schema: 2.0.0
+---
+
+# Test-AzDataProtectionBackupVaultNameAvailability
```
Comment (hiaga) [inline src/DataProtection/docs/Test-AzDataProtectionBackupVaultNameAvailability.md:8]: remove this file

File: src/DataProtection/docs/Test-AzDataProtectionFeatureSupport.md [hunk 18]
```diff
@@ -0,0 +1,232 @@
+---
+external help file:
+Module Name: Az.DataProtection
+online version: https://docs.microsoft.com/powershell/module/az.dataprotection/test-azdataprotectionfeaturesupport
+schema: 2.0.0
+---
+
```
Comment (hiaga) [inline src/DataProtection/docs/Test-AzDataProtectionFeatureSupport.md:7]: remove this file

File: src/DataProtection/examples/Set-AzDataProtectionMSIPermission.md [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+
+```
+The above command is used to assign permissions to the backup vault "Vaultname" under resource group "VaultRG" at the "Rresource Group" scope of the disk.
+
+
+
+### Example 2: Grant Permissions for Azure Blobs
+```powershell
+PS C:\> Set-AzDataProtectionMSIPermission -BackupInstance $instance -VaultResourceGroup "VaultRG"" -VaultName "Vaultname"" -PermissionsScope "ResourceGroup"
+```
```
Comment (hiaga) [inline src/DataProtection/examples/Set-AzDataProtectionMSIPermission.md:13]: same as above

File: src/DataProtection/examples/Test-AzDataProtectionBackupInstanceReadiness.md [hunk 20]
```diff
@@ -0,0 +1,22 @@
+### Example 1: {{ Add title here }}
+```powershell
+{{ Add code here }}
+```
+
+```output
+{{ Add output here }}
```
Comment (hiaga) [inline src/DataProtection/examples/Test-AzDataProtectionBackupInstanceReadiness.md:7]: add example

File: src/DataProtection/examples/Test-AzDataProtectionBackupInstanceRestore.md [hunk 21]
```diff
@@ -0,0 +1,22 @@
+### Example 1: {{ Add title here }}
+```powershell
+{{ Add code here }}
+```
+
```
Comment (hiaga) [inline src/DataProtection/examples/Test-AzDataProtectionBackupInstanceRestore.md:5]: add one example

File: src/DataProtection/examples/Test-AzDataProtectionBackupVaultNameAvailability.md [hunk 22]
```diff
@@ -0,0 +1,22 @@
+### Example 1: {{ Add title here }}
+```powershell
+{{ Add code here }}
+```
+
```
Comment (hiaga) [inline src/DataProtection/examples/Test-AzDataProtectionBackupVaultNameAvailability.md:5]: remove file

File: src/DataProtection/examples/Test-AzDataProtectionFeatureSupport.md [hunk 23]
```diff
@@ -0,0 +1,22 @@
+### Example 1: {{ Add title here }}
+```powershell
+{{ Add code here }}
+```
+
+```output
```
Comment (hiaga) [inline src/DataProtection/examples/Test-AzDataProtectionFeatureSupport.md:6]: remove file

File: src/DataProtection/test/Set-AzDataProtectionMSIPermission.Tests.ps1 [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+        $policyDefn = Get-AzDataProtectionPolicyTemplate -DatasourceType AzureBlob
+        
+        New-AzDataProtectionBackupPolicy -ResourceGroupName $VaultRG -VaultName $TestBkpVault.Name -Name blobBkpPolicy -Policy $policyDefn
+        $blobBkpPol = Get-AzDataProtectionBackupPolicy -ResourceGroupName $VaultRG -VaultName $TestBkpVault.Name -Name $BlobPolicyName
+
+        $instance = Initialize-AzDataProtectionBackupInstance -DatasourceType AzureBlob -DatasourceLocation $TestBkpvault.Location -PolicyId $blobBkpPol[0].Id -DatasourceId $BlobId
+        
+        #Validation should fail
+        try
+        {Test-AzDataProtectionBackupInstanceReadiness -ResourceGroupName $VaultRG -VaultName $TestBkpVault.Name -BackupInstance  $instance[0].Property}
```
Comment (hiaga) [inline src/DataProtection/test/Set-AzDataProtectionMSIPermission.Tests.ps1:72]: again make sure the validation fails with permission issue

File: src/DataProtection/test/env.json [hunk 25]
```diff
@@ -1,69 +1,85 @@
...
-    "FileNamePrefix": "oss-pstest-restoreasfiles",
-    "OssDbId": "/subscriptions/38304e13-357e-405e-9e9a-220351dcce8c/resourceGroups/hiagarg/providers/Microsoft.DBforPostgreSQL/servers/oss-pstest-server/databases/postgres",
-    "OssServerName": "oss-pstest-server",
-    "SecretURI": "https://oss-pstest-keyvault.vault.azure.net/secrets/oss-pstest-secret",
-    "ResourceGroupName": "oss-pstest-rg",
-    "PolicyName": "oss-pstest-policy"
+  "SubscriptionId": "62b829ee-7936-40c9-a1c9-47a93f9f3965",
+  "TestGrantPermission": {
+    "VaultRG": "testBkpVaultRG",
+    "ossrg": "/subscriptions/62b829ee-7936-40c9-a1c9-47a93f9f3965/resourcegroups/Ossrg",
```
Comment (hiaga) [inline src/DataProtection/test/env.json:18]: inconsistent casing

File: src/DataProtection/test/env.json [hunk 26]
```diff
@@ -1,69 +1,85 @@
...
+    "Diskrg": "/subscriptions/62b829ee-7936-40c9-a1c9-47a93f9f3965/resourceGroups/Diskrg",
+    "keyvaultId": "/subscriptions/62b829ee-7936-40c9-a1c9-47a93f9f3965/resourcegroups/Sqlrg/providers/Microsoft.KeyVault/vaults/rishitkeyvault3",
```
Comment (hiaga) [inline src/DataProtection/test/env.json:20]: same, correct the casing at all places in this file

File: src/DataProtection/custom/Helpers/BackupcenterHelper.ps1 [hunk 27]
```diff
@@ -30,6 +30,79 @@
...
+}
+
+
+function GetResourceNameFromArmId {
+	[Microsoft.Azure.PowerShell.Cmdlets.DataProtection.DoNotExportAttribute()]
+	param(
+		[Parameter(Mandatory=$true)]
+        [ValidateNotNullOrEmpty()]
+        [System.String]
+        $DatasourceId
```
Comment (hiaga) [inline src/DataProtection/custom/Helpers/BackupcenterHelper.ps1:76]: Instead of datesource Id - can we say just Id ?

File: src/DataProtection/custom/Helpers/BackupcenterHelper.ps1 [hunk 28]
```diff
@@ -30,6 +30,79 @@
...
+	)
+	
+	process{
+
+        $ResourceArray = $DataSourceId.Split("/")
```
Comment (hiaga) [inline src/DataProtection/custom/Helpers/BackupcenterHelper.ps1:81]: can we split into key value pairs ?

File: src/DataProtection/docs/Az.DataProtection.md [hunk 29]
```diff
@@ -24,7 +24,7 @@ Adds or removes schedule tag in an existing backup policy.
 Updates Backup schedule of an existing backup policy.
 
 ### [Find-AzDataProtectionRestorableTimeRange](Find-AzDataProtectionRestorableTimeRange.md)
-Finds the valid recovery point in time ranges for the restore.
+
```
Comment (hiaga) [inline src/DataProtection/docs/Az.DataProtection.md:27]: revert

File: src/DataProtection/docs/Az.DataProtection.md [hunk 30]
```diff
@@ -113,6 +116,18 @@ This operation will stop backups for backup instance
...
 
+### [Test-AzDataProtectionBackupInstanceReadiness](Test-AzDataProtectionBackupInstanceReadiness.md)
+Validate whether adhoc backup will be successful or not
+
+### [Test-AzDataProtectionBackupInstanceRestore](Test-AzDataProtectionBackupInstanceRestore.md)
+Validates if Restore can be triggered for a DataSource
+
+### [Test-AzDataProtectionBackupVaultNameAvailability](Test-AzDataProtectionBackupVaultNameAvailability.md)
+API to check for resource name availability
+
```
Comment (hiaga) [inline src/DataProtection/docs/Az.DataProtection.md:127]: remove from 125 to 130

File: src/DataProtection/docs/Find-AzDataProtectionRestorableTimeRange.md [hunk 31]
```diff
@@ -20,7 +21,7 @@ Find-AzDataProtectionRestorableTimeRange -BackupInstanceName <String> -ResourceG
 ```
 
 ## DESCRIPTION
-This cmdlet finds the valid restorable pointInTime ranges within which one can trigger restore for a given backup instance.
+
```
Comment (hiaga) [inline src/DataProtection/docs/Find-AzDataProtectionRestorableTimeRange.md:24]: revert this change

**Review by hiaga — CHANGES_REQUESTED** [review]

**Comment by hiaga** [comment]
@rishit-microsoft build is failing, please check

File: src/DataProtection/custom/Helpers/BackupcenterHelper.ps1 (hunk 27 above)
Comment (rishit-microsoft) [reply src/DataProtection/custom/Helpers/BackupcenterHelper.ps1:76]: Aditya's Document says the parameter name should be DatasourceId.So should I still go ahead and change it?

**Comment by hiaga** [comment]
@BethanyZhou : Can you please review this PR. I've currently removed the signature changes because it is causing the build to fail, Please suggest how to suppress the Static analysis failure

File: src/DataProtection/README.md [hunk 32]
```diff
@@ -92,6 +92,14 @@ directive:
       verb: Get
       subject: RecoveryPointList.*
     hide: true
+  - where:
```
Comment (BethanyZhou) [inline src/DataProtection/README.md:95]: please add branch tag to lock the code of https://github.com/Azure/azure-rest-api-specs/blob/main/specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2022-05-01/dataprotection.json in certain commit id.

File: src/DataProtection/custom/Helpers/BackupcenterHelper.ps1 [hunk 33]
```diff
@@ -30,6 +30,79 @@
 	}
 }
 
+function GetResourceGroupIdFromArmId {
+	[Microsoft.Azure.PowerShell.Cmdlets.DataProtection.DoNotExportAttribute()]
+	param(
+		[Parameter(Mandatory=$true)]
```
Comment (BethanyZhou) [inline src/DataProtection/custom/Helpers/BackupcenterHelper.ps1:36]: Could you correct these indents?

File: src/DataProtection/custom/Helpers/CommonHelpers.ps1 [hunk 34]
```diff
@@ -5,13 +5,13 @@ function GetDatasourceSetInfo
...
 		[ValidateNotNullOrEmpty()]
-		[Microsoft.Azure.PowerShell.Cmdlets.DataProtection.Models.Api20220401.IDatasource]
+		[Microsoft.Azure.PowerShell.Cmdlets.DataProtection.Models.Api20220501.IDatasource]
 		$DatasourceInfo
 	)
 
 	process 
 	{
-		$DataSourceSetInfo = [Microsoft.Azure.PowerShell.Cmdlets.DataProtection.Models.Api20220401.DatasourceSet]::new()
+		$DataSourceSetInfo = [Microsoft.Azure.PowerShell.Cmdlets.DataProtection.Models.Api20220501.DatasourceSet]::new()
```
Comment (BethanyZhou) [inline src/DataProtection/custom/Helpers/CommonHelpers.ps1:14]: Correct indent please.

File: src/DataProtection/docs/Set-AzDataProtectionMSIPermission.md [hunk 35]
```diff
@@ -0,0 +1,210 @@
...
+```
+
+## DESCRIPTION
+Grants required permissions to the backup vault to configure backup
+
+## EXAMPLES
+
+### Example 1: Grant Permissions for Azure Disks
+```powershell
+PS C:\> Set-AzDataProtectionMSIPermission -BackupInstance $instance -VaultResourceGroup "VaultRG"" -VaultName "Vaultname"" -PermissionsScope "ResourceGroup"
```
Comment (BethanyZhou) [inline src/DataProtection/docs/Set-AzDataProtectionMSIPermission.md:28]: Please remove PS C:\> from all examples to avoid breaking user's copy-paste-use experience.

File: src/DataProtection/docs/Set-AzDataProtectionMSIPermission.md [hunk 36]
```diff
@@ -0,0 +1,210 @@
...
+
+```
+
+The above command is used to assign permissions to the backup vault "Vaultname" under resource group "VaultRG" at the "Rresource Group" scope of the disk.
+
+### Example 2: Grant Permissions for Azure Blobs
+```powershell
+PS C:\> Set-AzDataProtectionMSIPermission -BackupInstance $instance -VaultResourceGroup "VaultRG"" -VaultName "Vaultname"" -PermissionsScope "Subscription"
+```
```
Comment (BethanyZhou) [inline src/DataProtection/docs/Set-AzDataProtectionMSIPermission.md:37]: could you add \```output {output}  \``` for each example?

File: src/DataProtection/docs/Set-AzDataProtectionMSIPermission.md [hunk 37]
```diff
@@ -0,0 +1,210 @@
...
+# Set-AzDataProtectionMSIPermission
+
+## SYNOPSIS
+Grants required permissions to the backup vault to configure backup
+
+## SYNTAX
+
+```
+Set-AzDataProtectionMSIPermission -BackupInstance <IBackupInstanceResource> -PermissionsScope <String>
+ -VaultName <String> -VaultResourceGroup <String> [-KeyVaultId <String>] [-Confirm] [-WhatIf]
```
Comment (BethanyZhou) [inline src/DataProtection/docs/Set-AzDataProtectionMSIPermission.md:17]: A resource group and a vault name is enough to identify a key vault, why we still need a KeyVaultId

File: src/DataProtection/docs/Test-AzDataProtectionBackupInstanceReadiness.md [hunk 38]
```diff
@@ -0,0 +1,323 @@
+---
+external help file:
+Module Name: Az.DataProtection
+online version: https://docs.microsoft.com/powershell/module/az.dataprotection/test-azdataprotectionbackupinstancereadiness
+schema: 2.0.0
+---
+
+# Test-AzDataProtectionBackupInstanceReadiness
```
Comment (BethanyZhou) [inline src/DataProtection/docs/Test-AzDataProtectionBackupInstanceReadiness.md:8]: For this command, I think we only need to keep variant ValidateExpanded

File: src/DataProtection/docs/Test-AzDataProtectionBackupInstanceRestore.md [hunk 39]
```diff
@@ -0,0 +1,344 @@
...
+```
+
+### Validate1
+```
+Test-AzDataProtectionBackupInstanceRestore -Name <String> -ResourceGroupName <String> -VaultName <String>
+ -Parameter <IValidateRestoreRequestObject> [-SubscriptionId <String>] [-DefaultProfile <PSObject>] [-AsJob]
+ [-NoWait] [-Confirm] [-WhatIf] [<CommonParameters>]
+```
+
+### ValidateExpanded1
```
Comment (BethanyZhou) [inline src/DataProtection/docs/Test-AzDataProtectionBackupInstanceRestore.md:29]: I think we only need to keep parameter set ValidateExpanded1

File: src/DataProtection/README.md (hunk 32 above)
Comment (hiaga) [reply src/DataProtection/README.md:95]: Done

File: src/DataProtection/docs/Set-AzDataProtectionMSIPermission.md (hunk 37 above)
Comment (hiaga) [reply src/DataProtection/docs/Set-AzDataProtectionMSIPermission.md:17]: We need this because **resource group and vault name correspond to backup vault**, whereas **KeyVaultId** is a parameter used in a specific case of granting permissions in Postgres server scenario

**Comment by BethanyZhou** [comment]
Comments are not resolved but conversations were marked as resolved. Did you forget to push commit here?

**Comment by isra-fel** [comment]
Kindly ping @rishit-microsoft

**Comment by isra-fel** [comment]
Closing the PR for lack of activity. Feel free to reopen to continue working on it. Thanks 😀

**Comment by hiaga** [comment]
@isra-fel  @BethanyZhou  apologies for leaving the PR inactive for quite sometime, we were focusing on some other PS deliverables in the meantime. This PR was almost ready, I'll upload the commit after resolving comments and will now work towards it's closure.

File: src/DataProtection/docs/Test-AzDataProtectionBackupInstanceReadiness.md (hunk 38 above)
Comment (hiaga) [reply src/DataProtection/docs/Test-AzDataProtectionBackupInstanceReadiness.md:8]: removed extra param sets

File: src/DataProtection/docs/Test-AzDataProtectionBackupInstanceRestore.md (hunk 39 above)
Comment (hiaga) [reply src/DataProtection/docs/Test-AzDataProtectionBackupInstanceRestore.md]: removed extra param sets

File: src/DataProtection/README.md [hunk 40]
```diff
@@ -36,7 +36,7 @@ This file contains the configuration for generating My API from the OpenAPI spec
 # it's the same options as command line options, just drop the double-dash!
 require:
   - $(this-folder)/../readme.azure.noprofile.md
-input-file: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2022-04-01/dataprotection.json
+input-file: https://github.com/Azure/azure-rest-api-specs/blob/c0d5296b483a5fe4de7851fcd45acde14e736574/specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2022-05-01/dataprotection.json
```
Comment (wyunchi-ms) [inline src/DataProtection/README.md:39]: Please use `branch` such as this: https://github.com/Azure/azure-powershell/blob/a501c63bc3daddbf5695c706a3043a537fea1bca/src/Databricks/README.md?plain=1#L51

File: src/DataProtection/docs/Set-AzDataProtectionMSIPermission.md (hunk 37 above)
Comment (wyunchi-ms) [reply src/DataProtection/docs/Set-AzDataProtectionMSIPermission.md:17]: So there will be two KeyVault in `Postgres server scemario`?

**Comment by hiaga** [comment]
So there will be two KeyVault in Postgres server scemario?

no, only one key vault and one backup vault. backup vault is by default used in all data protection commands 

 -VaultName <String> -VaultResourceGroup <String>  - this inputs the backup vault for all scenarios 
 
 [-KeyVaultId <String>] - this inputs the key vault for Postgres server scenario

File: src/DataProtection/README.md (hunk 40 above)
Comment (hiaga) [reply src/DataProtection/README.md:39]: sure, will do

File: src/DataProtection/docs/Set-AzDataProtectionMSIPermission.md (hunk 37 above)
Comment (hiaga) [reply src/DataProtection/docs/Set-AzDataProtectionMSIPermission.md:17]: no, only one key vault and one backup vault. backup vault is by default used in all data protection commands
-VaultName -VaultResourceGroup - this inputs the backup vault for all scenarios
[-KeyVaultId ] - this inputs the key vault for Postgres server scenario

**Review by wyunchi-ms — APPROVED** [review]
